architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,495 @@
+- (non-goal): Support all existing API-design constructs in declarative user-defined types. 
+	- It's our goal to simplify the API design guidance for users. We should be open to rethinking existing practices if they cause us or users difficulty.
+- (non-goal): Support compatibility with existing APIs/features as we migrate existing Radius functionality. 
+	- It's our goal to simplify the experience for users. We should be open to rethinking existing features and designs if it is overall better. 
+- (out of scope): Building an SDK for writing a resource provider microservices. Our priority is to simplify the experience for authoring resource types declaratively. We will wait for feedback before creating comprehensive support to author a resource provider microservice.
+- (out of scope): Supporting resource types for use cases other than applications. Our priority is to simplify the experience for authoring resource types that will be used in applications. 
+  - eg: Extensibility for other planes like Azure. (It doesn't make sense).
+  - eg: Resources in the Radius plane that use scopes other than resource group. (Not a top priority, can be added in the future).
+  
+### User scenarios (optional)
+
+The scenario document for user-defined types is in progress [here](https://github.com/radius-project/design-notes/pull/58). Please refer to the scenarios defined in that doc.
+
+## User Experience (if applicable)
+
+Proposed user-experience changes are defined in the [scenario doc](https://github.com/radius-project/design-notes/pull/58).
+
+Here's a abstract summary. 
+
+Jenny is a platform engineer implementing support for PostgreSQL in Radius. Her E2E experience authoring a declarative user-defined type looks like this:
+
+**Authoring**
+
+1. Choose a name for the new type. She chooses `Contoso.Example/postgreSQLDatabases`.
+2. Design a schema for the new type. She's inspired by the design of `Applications.Datastores/sqlDatabases` (MS SQL Server) and chooses something similar. Her new type will return fields like `database`, `port`, `host`, `username`, `password`, and `url`. 
+3. Author a manifest for the new resource provider `Contoso.Example`. She needs to make a few choices at this stage:
+   1. Choose an API version, she picks `2024-08-1`. 
+   2. Implement an RP imperatively in code, or declaratively with support for recipes. She wants to use recipes and the declarative approach.
+4. Use the tools + manifest to generate:
+   1.  A Bicep extension (like the `radius` or `aws` Bicep extension)
+   2.  A "cooked" version of the manifest that can be used with `rad resourceprovider create ...`. 
+
+*At this point she's done authoring code/assets for the new type and resource provider.*
+
+**Using**
+
+1. Use `rad resourceprovider create ...` to register the resource provider with Radius.
```

Comment:
**RP** is **Contoso.Example** and resource type is **postgreSQLDatabases**, right? Do we have another command for creating the resource type?

Reply:
`Contoso.Example` is the Resource Provider and Resource Provider namespace.

`postgreSQLDatabases` is a partially-qualified resource type.

`Contoso.Example/postgreSQLDatabases` is a fully-qualified resource type.